### ISSUE TITLE
Mask "unconstrained" pixels in 3D background model

### ISSUE THREAD

**Issue by lmohrmann** [gammapy/gammapy#2432]
This issue is to start a discussion about the treatment of spatial pixels in 3D background models for which no reasonable background estimate is possible.

I've just opened an issue with the data format specs, see open-gamma-ray-astro/gamma-astro-data-formats#135, proposing to set such pixels to `NaN`.

Gammapy will need to be updated to properly treat these pixels (mask them in a fit, ignore them in stacking, ...).

It would be very nice to have a working solution for this in v0.15, as we would like to mark pixels in this (or a similar) way in new background models that we produce for HESS.

Opinions?

@cdeil @registerrier @adonath

**Comment by cdeil** [comment]
I think putting `NaN` into the DL3 IRFs is problematic, and instead we should leave the bad values (or in some cases put zeros) for now, and in the near future (starting next year, not now?) improve by giving quality information separately (e.g. as a mask or weights image), meaning that science tools can gradually add support and use that information if present.

The problem with `NaN` is that you have to make a hard decision into "good" and "bad", where really there's all shades of gray with IRF quality, and different science cases require different choices. E.g. at very high energy you won't have any background estimate. Do you put `NaN`, and should science tools forbid the analysis of these energies? What if the source is very bright and background doesn't matter to get a spectrum? Should science tools just replace `NaN` with zero in that case? Then I think just putting zero from the start is better (no changes to science tools needed).

At low energies having `NaN` is also problematic, because the EDISP can shift those `NaN` to pretty high energies, i.e. you don't get a spectrum near the energy threshold at all? Again, I think it's better to keep "bad" EDISP and PSF at low energies, and then because AEFF is very low sometimes they don't matter. Sometimes you do get completely incorrect results near the energy threshold. But if you blank out IRFs at low energies with `NaN` at the DL3 level, you prevent the option of people to explore and assess analysis results near the energy threshold at all.

I don't think the Fermi IRFs have `NaN`, do they?
And in other domains (optical, X-ray), I think weighting for quality information is common (see e.g. [here](https://sextractor.readthedocs.io/en/latest/Weighting.html)).

@lmohrmann - I share the concern of silently getting bad analysis results, and agree that IRF quality information should be added in DL3 and used in Gammapy. But I would suggest to basically keep things as-is for now in your background IRF production, to help do the ton of work we need for v1.0, and then next year together with CTA to re-tackle the IRF work, i.e. defer this issue to v1.1 now.

**Comment by cdeil** [comment]
One more comment: if we spend a few more weeks and clean up gammapy.irf (see #776), adding e.g. an optional quality mask will be simpler. I think we already have such a mask on the datasets, i.e. you could then export it as an extra column in the DL3 IRFs, and if present start using it in Gammapy, i.e. it's a way forward that doesn't require agreement by all stakeholders for DL3, it's a backward-compatible addition, which then once CTA has made a choice (possibly in 1-2-3 years, not soon) Gammapy code can adapt to that solution.

**Comment by lmohrmann** [comment]
@cdeil 
I agree that in the long term, there will need be to a proper IRF validity information as part of the DL3 data format (Karl already commented on this in open-gamma-ray-astro/gamma-astro-data-formats#135). However, I would very much like to have a working solution for v0.15, not in v1.0 or later.

On this time scale, I think the reasonable thing to do is to mark pixels as "bad" by using some to-be-agreed-upon value. I would prefer `NaN`, because that is very explicit, but if there is a preference for, say, putting `0`, that's also fine with me. In any case, we need to make sure that these pixels are properly treated in the analysis (i.e., not used).

Note that I'm only talking about the 3D background model for the moment - I do not plan to export other IRFs like EDISP or PSF with `NaN` values any time soon.

@adonath @registerrier What are your opinions on this? If we go for marking bad pixels with `0`, are there still changes that we need to apply in Gammapy to make sure such pixels are masked?

**Comment by cdeil** [comment]
@lmohrmann - If you put NaN in the background IRF, what would the default behaviour in Gammapy be - how should it handle those values at which point in the analysis?

**Comment by lmohrmann** [comment]
@cdeil - I don't have a good overview over all cases that need to be considered. Two obvious cases are
- fitting: the pixels should be masked
- stacking: the pixels should be ignored

I admit that I do not have a full solution for this yet, and there might well be cases where it's difficult to deal with `NaN`s. On the other hand, I think essentially the same would apply for other values that we put to mark invalid pixels - as far as I understand there is no support yet for marking pixels as invalid, by whatever mechanism, so this would need to be added in any case.

**Comment by cdeil** [comment]
What we have at the moment is support in Gammapy for the safe energy and safe max offset parameters. Reviewing how exactly those are used in 1D & 3D joint and stacked analysis, fixing bugs if needed, and documenting in a short paragraph the current behaviour would be a very valuable contribution. Details matter: does each IRF have their own safe range, or do we only use the values from the AEFF? In stacking, we probably just "do nothing" and just sum, meaning the stacked datasets extends to the lowest energy from any run? How are bin centers or edges used when applying safe energy ranges? This I consider really important for v1.0, should be done first.

The next most important thing, that Johannes and I planned to add to the FITS exporters but then stopped working on it, was to have an offset-dependent safe energy cut, as an extra BINTABLE, that if present would be used instead of the single safe energy value for all offsets. That would lead to better results (recovered exposure) at low energy. You could just skip that step and directly export a QUALITY mask that matches the IRF array, where you put 1 to keep and 0 to ignore.

Like I said: I think putting IRF quality information separately, instead of blanking out via NaN, is better, because when blanking out you destroy the IRF information in the FITS and make studies how the safe energy thresholds affect analysis results impossible. Really different science use cases require different choices, e.g. for GRB detection you want extremely loose cuts, for "precision" measurements like Crab position or spectrum you want strict cuts. So in the future it could be envisioned that there's multiple QUALITY infos, although probably one would just keep it simple and release multiple FITS "configs".

Specifically for background, I think you're probably putting 0 now where you have no background information (it's what naturally comes out when producing background IRFs as smoothed counts from off data), and that's great. Releasing those background models to the HESS collaboration to let people start doing analyses should be *much* higher priority than to improve quality handling  further beyond the safe energy and max offset cut that we have now (what HESS has done for the past decade). It just means people have to watch out for analysis issues near the safe energy cut and at the edge of the FOV, but that's a task that will always remain, only the default behaviour can be improved by putting better IRFs or QUALITY masks. Of course, priorities and goals are an opinion, that's just my 2 cents. Note that e.g. we never checked if EDISP is properly treated in Gammapy at all, I would consider that another high-priority task that ideally we'd do before v1.0 (see https://github.com/gammapy/gammapy/issues/1538#issuecomment-538113964).

@lmohrmann - It's great if you take on this topic, do some prototyping, then update the open spec to allow for more extensive QUALITY information, than implement handling of that in Gammapy. That'll be super useful, and CTA might just adopt whatever you develop, or they might develop something different in a few years and then we adapt. I just think it's a very bad time for us from the Gammapy project to put effort on this now, because we need to polish and wrap up for v1.0 and just have 2 months left. So if you're willing to help out with that, e.g. by checking and documenting current behaviour as mentioned above, or scripting and checking if we can reproduce Gammapy results from the HESS FITS validation paper, that would be super, and then you could put the IRF v2.0 with better QUALITY in 2020. Having the scripted HESS validation analyses as baseline would be the perfect test case to put more extensive QUALITY information and re-run the analyses, and see how they change e.g. at low energy.

**Comment by lmohrmann** [comment]
@cdeil - I never wrote that I propose to have a fully working solution for IRF quality in v1.0, nor did I commit to start prototyping such a thing. My main motivation here was to start a discussion about whether it would be feasible to have the possibility to handle invalid pixels in the 3D background model (not any of the other IRFs!) already in v0.15.

To explain: we're in the process of creating new background models for internal use in HESS, and I'd like to know how to deal with pixels where we cannot make a reasonable prediction of the background rate. These new background models are the next important step for establishing Gammapy as an analysis tool within HESS, and this is why I am eager to have a (intermediate) solution for the next release.

@adonath - I can understand that you hesitate to implement something preliminary now, and remove it again in the future (not clear to me how near that future is though, getting to a full solution for IRF quality is likely on a time scale of several months...?). But the minimal action that you propose sounds like a reasonable plan to me.

So, to wrap up: the idea would be that we mark invalid pixels with `0` and we make a few minimal changes in Gammapy such that pixels with zero predicted background rate never get assigned a rate >0. Masking of these pixels could then be done "by hand" by the user. Would that be OK for everyone?

**Comment by adonath** [comment]
Thanks @lmohrmann, this would be my preferred compromise.

I just remembered, that in the code for computation of TS maps, I introduced exactly the same pattern: https://github.com/gammapy/gammapy/blob/master/gammapy/detect/test_statistics.py#L209
Pixels with `exposure > 0`, but `background = 0` are excluded from the likelihood computation.

The two action items for Gammapy are:
- Adapt `LogScale` to conserve zeros on input (possibly via `-np.inf`, need to check though...)
- Probably show how to compute the custom mask for unconstrained background pixels somewhere in the tutorial with a short explanation...

**Comment by adonath** [comment]
One more note: this could be now implemented as an additional method for the `SafeMaskMaker` for convenience.

**Comment by adonath** [comment]
@lmohrmann What is the current status on this issue here? I'm asking because @QRemy proposed a method to interpolate zero bins, based on neighbouring bins in #3219. I think this is technically a fair solution, however I'm not 100% happy to include this into the Gammnapy code base, because I think it's not the right place. I think it should happen on IRF production instead, where smoothing etc. happens as well. Have you considered such a method already?

**Comment by lmohrmann** [comment]
@adonath We have not implemented such a thing yet, no. I do agree in principle that ideally this is something to be solved in the IRF production, not the science tools. However: there's always the possibility that Gammapy is being handed an IRF that has not been properly prepared/smoothed/interpolated, and I think it's a good strategy to try to ensure that nothing really bad happens in this case. It's not so easy for the user to identify a bad extrapolation in just a few bins of the background cube, so this can lead to a lot of confusion.

For my understanding: has the compromise discussed above ever been implemented? (i.e. are pixels with background rate 0 guaranteed to remain at 0 also after interpolation?)

**Comment by QRemy** [comment]
The existing solution will work for interpolation but not always for extrapolation,  for example if the last bin is well defined but the one before is zero then the extrapolation beyond the last bin will result in aberrantly high values.

**Comment by adonath** [comment]
Thanks @lmohrmann for the quick answer! In the dev call we decided to introduce a little helper function, that fixes the "missing values", that can be optionally used in the `MapDatasetMaker`. The hope is, that this workaround can be deprecated  again, once the IRF production has stabilised.

**Comment by lmohrmann** [comment]
Sounds good! :+1:

**Comment by adonath** [comment]
Should be resolved by #3416 and #3418.
